File: 0641. Design Circular Deque/Solution.py
class MyCircularDeque:

    def __init__(self, k: int):
        """
        Initialize your data structure here. Set the size of the deque to be k.
        """
        self.k = k
        self.element = []
        

    def insertFront(self, value: int) -> bool:
        """
        Adds an item at the front of Deque. Return true if the operation is successful.
        """
        if len(self.element) < self.k:
            self.element.insert(0, value)
            return True
        else:
            return False
        

    def insertLast(self, value: int) -> bool:
        """
        Adds an item at the rear of Deque. Return true if the operation is successful.
        """
        if len(self.element) < self.k:
            self.element.append(value)
            return True
        else:
            return False
        

    def deleteFront(self) -> bool:
        """
        Deletes an item from the front of Deque. Return true if the operation is successful.
        """
        try:
            self.element.pop(0)
            return True
        except:
            return False
        

    def deleteLast(self) -> bool:
        """
        Deletes an item from the rear of Deque. Return true if the operation is successful.
        """
        try:
            self.element.pop()
            return True
        except:
            return False
        

    def getFront(self) -> int:
        """
        Get the front item from the deque.
        """
        try:
            return self.element[0]
        except:
            return -1
        

    def getRear(self) -> int:
        """
        Get the last item from the deque.
        """
        try:
            return self.element[-1]
        except:
            return -1
        

    def isEmpty(self) -> bool:
        """
        Checks whether the circular deque is empty or not.
        """
        return len(self.element) == 0
        

    def isFull(self) -> bool:
        """
        Checks whether the circular deque is full or not.
        """
        return len(self.element) == self.k
        


# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()

File: 0641. Design Circular Deque/Solution.cpp
class MyCircularDeque {
public:
    int k;
    std::vector<int> elements;
    
    /** Initialize your data structure here. Set the size of the deque to be k. */
    MyCircularDeque(int k) {
        this->k = k;
        this->elements.reserve(k);
    }
    
    /** Adds an item at the front of Deque. Return true if the operation is successful. */
    bool insertFront(int value) {
        if (this->elements.size() < this->k)
        {
            this->elements.insert(elements.begin(), value);
            return true;
        }
        else
        {
            return false;
        }
    }
    
    /** Adds an item at the rear of Deque. Return true if the operation is successful. */
    bool insertLast(int value) {
        if (this->elements.size() < this->k)        
        {
            this->elements.push_back(value);
            return true;
        }
        else
        {
            return false;
        }
    }
    
    /** Deletes an item from the front of Deque. Return true if the operation is successful. */
    bool deleteFront() {
        if (this->elements.empty())
        {
            return false;
        }
        else
        {
            this->elements.erase(this->elements.begin());
            return true;
        }
    }
    
    /** Deletes an item from the rear of Deque. Return true if the operation is successful. */
    bool deleteLast() {
        if (this->elements.empty())
        {
            return false;
        }
        else
        {
            this->elements.pop_back();
            return true;
        }        
    }
    
    /** Get the front item from the deque. */
    int getFront() {
        if (this->elements.empty())
        {
            return -1;
        }
        else
        {
            return this->elements[0];
        }
    }
    
    /** Get the last item from the deque. */
    int getRear() {
        if (this->elements.empty())
        {
            return -1;
        }
        else
        {
            return this->elements[this->elements.size()-1];
        }
    }
    
    /** Checks whether the circular deque is empty or not. */
    bool isEmpty() {
        return this->elements.empty();
    }
    
    /** Checks whether the circular deque is full or not. */
    bool isFull() {
        return this->elements.size() == this->k;
    }
};

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque* obj = new MyCircularDeque(k);
 * bool param_1 = obj->insertFront(value);
 * bool param_2 = obj->insertLast(value);
 * bool param_3 = obj->deleteFront();
 * bool param_4 = obj->deleteLast();
 * int param_5 = obj->getFront();
 * int param_6 = obj->getRear();
 * bool param_7 = obj->isEmpty();
 * bool param_8 = obj->isFull();
 */
